src/main/kotlin/anttracker/PageOf.kt

```diff
@@ -0,0 +1,119 @@
+        for (record in records) {
+            printRecord(record)
+        }
+        if (!lastPage()) {
+            println("<Enter> for ${countRemainingRecords()} more.")
+        }
+    }
+
+    // -------------------------------------------------------------------------------
+    // Queries to DB and pulls a page of records (max 20) into the records MutableList.
+    // ---
+    fun loadRecords() {
+        records.clear()
+        transaction {
+            val queryOutput = queryToDB()
+            if (queryOutput != null) {
+                queryOutput.forEach { record ->
+                    records.add(record)
+                }
+            } else {
+                throw Exception("PageOf: Attempted to load from null query output.")
+            }
+        }
+    }
+
+    // -------------------------------------------------------------------------------
+    // Returns the size/length of the records MutableList.
+    // ---
+    fun recordsSize(): Int = records.size
+
+    // -------------------------------------------------------------------------------
+    // Getter for elements in the records MutableList.
+    // ---
+    fun getContentAt(index: Int): IntEntity = records[index]
+
+    // -------------------------------------------------------------------------------
+    // Increments page number and loads associated DB records into records MutableList
+    // Throws an error if attempting to load beyond the last page.
+    // ---
+    fun loadNextPage() {
+        if (lastPage()) {
+            throw Exception("PageOf: Already reached last page.")
+        }
+        pagenum++
+        loadRecords()
+    }
+
+    // -------------------------------------------------------------------------------
+    // TRUE if current page >= than the last page.
+    // ---
+    fun lastPage(): Boolean = pagenum >= lastPageNum
+
+    // -------------------------------------------------------------------------------
+    // Returns the number of yet-to-be-displayed records after the current page.
+    // ---
+    fun countRemainingRecords(): Int = getQueryRecordCount() - (queryLimit * (pagenum + 1))
+
+    // -------------------------------------------------------------------------------
+    // Used in init/ctor block to calculate the last page number.
+    // As we are 0-indexing the page numbers, we subtract 1 at the end.
+    // ---
+    protected fun initLastPageNum() {
+        lastPageNum =
+            ceil(
+                (getQueryRecordCount().toDouble() / queryLimit.toDouble()),
+            ).toInt() - 1
+    }
+
+    // -------------------------------------------------------------------------------
+    // Returns the total number of records/rows of the query being paginated.
+    // Used for calculating the number of pages the query needs.
+    // ---
+    private fun getQueryRecordCount(): Int {
+        var count: Int = -1
+        transaction {
+            val queryOutput: SizedIterable<IntEntity>? = queryToDB()
```

Comment:
If we call `queryToDB`, it will always return at most one page of records. That means the count will be 20 or less.

I think we should do the count directly, like this. 

```kotlin
 (Releases innerJoin Products)
        .select(Releases.columns)
        .where { Products.name eq productName }
        .count()
```

Reply:
I've abstracted the query to a getQuery() function that returns the Query function. This now allows fetchPageOfRecords to grab a page from the query, while initLastLineNum() can calc the number of pages from the entire query. 
So subclasses now just need to implement getQuery(), while fetchPageOfRecords() is now defined in the abstract base class.